### PR TITLE
NOISSUE - Trace MQTT forward operations

### PR DESCRIPTION
**Pull request by SammyOina**
Pull request title should be `MF-XXX - description` or `NOISSUE - description` where XXX is ID of issue that this PR relate to.
Please review the [CONTRIBUTING.md](https://github.com/mainflux/mainflux/blob/master/CONTRIBUTING.md) file for detailed contributing guidelines.

### What does this do?
Trace MQTT forward operations

### Which issue(s) does this PR fix/relate to?
There is no such issue.

### List any changes that modify/break current functionality
None

### Have you included tests for your changes?
Ni

### Did you document any new/modified functionality?
No

### CODE REVIEW

File: mqtt/forwarder.go [hunk 1]
```diff
@@ -20,7 +21,7 @@ const (
 type Forwarder interface {
 	// Forward subscribes to the Subscriber and
 	// publishes messages using provided Publisher.
-	Forward(id string, sub messaging.Subscriber, pub messaging.Publisher) error
+	Forward(ctx context.Context, id string, sub messaging.Subscriber, pub messaging.Publisher) error
```
Comment (arvindh123) [inline mqtt/forwarder.go]: We can replace the variable `ctx` with `_`, since it not used inside the function

File: mqtt/forwarder.go [hunk 2]
```diff
@@ -36,7 +37,7 @@ func NewForwarder(topic string, logger log.Logger) Forwarder {
 	}
 }
 
-func (f forwarder) Forward(id string, sub messaging.Subscriber, pub messaging.Publisher) error {
+func (f forwarder) Forward(ctx context.Context, id string, sub messaging.Subscriber, pub messaging.Publisher) error {
```
Comment (arvindh123) [inline mqtt/forwarder.go]: We can replace the variable `ctx` with `_`, since it not used inside the function

**Comment by dborovcanin** [comment]
@SammyOina  The goal of this initiative is not to add a tracer to the `Forward` method but to improve tracing on the MQTT Proxy side. This implies some changes in the Handler interface in [mProxy](https://github.com/mainflux/mproxy/blob/master/pkg/session/handler.go). We need context there anyways, so it's an excellent opportunity to add it now.

File: go.mod [hunk 3]
```diff
@@ -2,6 +2,8 @@ module github.com/mainflux/mainflux
 
 go 1.19
 
+replace github.com/mainflux/mproxy => /home/sammy/Documents/mproxy
```
Comment (rodneyosodo) [inline go.mod]: The PR was merged so you can update the version of mproxy

File: mqtt/tracing/forwarder.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+var _ mqtt.Forwarder = (*forwarderMiddleware)(nil)
+
+type forwarderMiddleware struct {
+	topic     string
+	forwarder mqtt.Forwarder
+	tracer    opentracing.Tracer
+}
+
+// New creates new mqtt forwarder tracing middleware.
+func New(tracer opentracing.Tracer, forwarder mqtt.Forwarder) mqtt.Forwarder {
```
Comment (rodneyosodo) [inline mqtt/tracing/forwarder.go]:
```suggestion
func New(topic string, tracer opentracing.Tracer, forwarder mqtt.Forwarder) mqtt.Forwarder {
```

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: go.mod (hunk 3 above)
Comment (SammyOina) [reply go.mod]: https://github.com/mainflux/mproxy/pull/36

File: mqtt/tracing/forwarder.go (hunk 4 above)
Comment (SammyOina) [reply mqtt/tracing/forwarder.go]: https://github.com/mainflux/mainflux/pull/1739#discussion_r1171182363

File: mqtt/tracing/forwarder.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+		tracer:    tracer,
+	}
+}
+
+// Forward traces mqtt forward operations
+func (fm *forwarderMiddleware) Forward(ctx context.Context, id string, sub messaging.Subscriber, pub messaging.Publisher) error {
+	span, ctx := opentracing.StartSpanFromContextWithTracer(ctx, fm.tracer, forwardOP)
+	defer span.Finish()
+	span.SetTag("subscriber", id)
+	span.SetTag("topic", fm.topic)
```
Comment (rodneyosodo) [inline mqtt/tracing/forwarder.go:35]: This means `fm.topic` will be empty

File: mqtt/tracing/forwarder.go (hunk 4 above)
Comment (rodneyosodo) [reply mqtt/tracing/forwarder.go]: https://github.com/mainflux/mainflux/pull/1775#discussion_r1180184888

File: mqtt/tracing/handler.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+	span, ctx := opentracing.StartSpanFromContextWithTracer(ctx, h.tracer, authConnectOP)
+	defer span.Finish()
+	return h.handler.AuthConnect(ctx)
+}
+
+// AuthPublish traces auth publish operations.
+func (h *handlerMiddleware) AuthPublish(ctx context.Context, topic *string, payload *[]byte) error {
+	span, ctx := opentracing.StartSpanFromContextWithTracer(ctx, h.tracer, authPublishOP)
+	defer span.Finish()
+	return h.handler.AuthPublish(ctx, topic, payload)
```
Comment (dborovcanin) [inline mqtt/tracing/handler.go:47]: Should we enrich the context with the topic and, maybe, publisher ID?  The same goes for other operations.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Comment by SammyOina** [comment]
image
image

**Review by dborovcanin — APPROVED** [review]

**Review by drasko — APPROVED** [review]
LGTM

**Pull request merged**